File: Leetcode/Minimum Cost For Tickets/Minimum Cost For Tickets.java
class Solution {

    public int mincostTickets(int[] days, int[] costs) {
      // Finding the last travel day
    int lastDay = days[days.length - 1]; 
    // dp array will contain the answers for i days
    int[] dp = new int[lastDay + 1]; 
    boolean[] isTravelDay = new boolean[lastDay + 1];
    //marking all the day that we are travelling
    for(int day : days) isTravelDay[day] = true;
    
        //Traversing through all days till lastDay
    for(int i = 1; i <= lastDay; i++) {
        // If not a travel day, we will not buy a ticket
        if(!isTravelDay[i]) { 
            dp[i] = dp[i - 1];
            continue;
        }
       // Checking for all the cases to minimum amount
        dp[i] = costs[0] + dp[i - 1]; // 1-day
        dp[i] = Math.min(costs[1] + dp[Math.max(i - 7, 0)], dp[i]); // 7-day
        dp[i] = Math.min(costs[2] + dp[Math.max(i - 30, 0)], dp[i]); // 30-day
    }
        //Returning the answer for given input
    return dp[lastDay]; 
    }
}

File: Leetcode/Minimum Cost For Tickets/Minimum_cost_for_tickets.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    int mincostTickets(vector<int>& days, vector<int>& costs)
     {
        //vector for storing cost per day
        vector<int> dp(366, INT_MAX);
        for (int day : days)
         {    dp[day] = 0;}
        
        //setting it to 0
        dp[0] = 0;

        //loop for finding minimun cost
        for (int i = 1; i < 366; ++i) 
        {
            // dp[i] means min cost for day i
            if (dp[i] == INT_MAX)
               {  dp[i] = dp[i - 1];}
            else {
                //checking conditions by comparing 
                int cur = dp[i - 1] + costs[0];
                cur = min(cur, costs[1] + dp[max(0, i - 7)]);
                cur = min(cur, costs[2] + dp[max(0, i - 30)]);
                dp[i] = cur;
            }
        }

        //returning minimum cost for tickets
        return dp[days[days.size() - 1]];
     }
};
